boa/src/syntax/lexer/number.rs

```diff
@@ -0,0 +1,349 @@
+                        kind = NumericKind::Integer(8);
+                    }
+                    'b' | 'B' => {
+                        // Remove the initial '0' from buffer.
+                        cursor.next();
+                        buf.pop();
+
+                        // BinaryIntegerLiteral
+                        kind = NumericKind::Integer(2);
+                    }
+                    'n' => {
+                        cursor.next();
+
+                        // DecimalBigIntegerLiteral '0n'
+                        return Ok(Token::new(
+                            TokenKind::NumericLiteral(Numeric::BigInt(0.into())),
+                            Span::new(start_pos, cursor.pos()),
+                        ));
+                    }
+                    ch => {
+                        if ch.is_digit(8) {
+                            // LegacyOctalIntegerLiteral
+                            if self.strict_mode {
+                                // LegacyOctalIntegerLiteral is forbidden with strict mode true.
+                                return Err(Error::syntax(
+                                    "Implicit octal literals are not allowed in strict mode.",
+                                ));
+                            } else {
+                                // Remove the initial '0' from buffer.
+                                buf.pop();
+
+                                buf.push(cursor.next().expect("'0' character vanished")?);
+
+                                kind = NumericKind::Integer(8);
+                            }
+                        } else if ch.is_digit(10) {
+                            // Indicates a numerical digit comes after then 0 but it isn't an octal digit
+                            // so therefore this must be a number with an unneeded leading 0. This is
+                            // forbidden in strict mode.
+                            if self.strict_mode {
+                                return Err(Error::syntax(
+                                    "Leading 0's are not allowed in strict mode.",
+                                ));
+                            } else {
+                                buf.push(cursor.next().expect("Number digit vanished")?);
+                            }
+                        } // Else indicates that the symbol is a non-number.
+                    }
+                }
+            } else {
+                // DecimalLiteral lexing.
+                // Indicates that the number is just a single 0.
+                return Ok(Token::new(
+                    TokenKind::NumericLiteral(Numeric::Integer(0)),
+                    Span::new(start_pos, cursor.pos()),
+                ));
+            }
+        }
+
+        // Consume digits until a non-digit character is encountered or all the characters are consumed.
+        cursor.take_until_pred(&mut buf, &|c: char| c.is_digit(kind.base()))?;
+
+        let exp_str = &mut String::new();
```

Comment:
why do we need `exp_str` we can just append to `buf`, right? and use `buf`

Reply:
from_str_radix can't handle exponent notation in the form 10E2 as it cannot differentiate between that and hex so I think they need to be separate

Reply:
> `from_str_radix` can't handle exponent notation in the form 10E2 as it cannot differentiate between that and hex so I think they need to be separate

why not do something like for integer case:
```
if let Ok(integer) = i32::from_str_radix(&buf, base) {
	// return integer;
} else {
	let b = f64::from(base);
    let mut result = 0.0_f64;
    for c in buf.chars() {
        let digit = f64::from(c.to_digit(base as u32));
    	result = result * b + digit;
    }
	// return result
}
```

Reply:
Will this mean that if we have an Exponent the result is always a rational?

Reply:
Yes. It will. we could have a case to see if it fits in a `i32` like:
```rust
if let Ok(integer) = i32::from_str_radix(&buf, base) {
	// return integer;
} else {
	let b = f64::from(base);
    let mut result = 0.0_f64;
    for c in buf.chars() {
        let digit = f64::from(c.to_digit(base as u32));
    	result = result * b + digit;
    }
	if let Ok(integer) = i32::try_from(result) {
		// return integer;
	}
	// return result 
}
```

Reply:
We decided previously (earlier comment) that we will try and keep the number as an integer if possible even if an exponential is used for performance reasons

Reply:
Well this is the Integer case and we already see if it fits in a `i32` here:
```rust
if let Ok(integer) = i32::from_str_radix(&buf, base) {
	// return integer;
```

and if it does not fit we retrun a rational:
```rust
} else {
	let b = f64::from(base);
    let mut result = 0.0_f64;
    for c in buf.chars() {
        let digit = f64::from(c.to_digit(base as u32));
    	result = result * b + digit;
    }
	// return result
}
```

We already try to fit it in an integer (`i32`), so I'm confused about the exponential part you are saying :confused:

Reply:
from_str_radix doesn't support e/E so if you wrote 10E2 then you could treat it as a rational and get 1000.0 or treat the 10 and the 2 separately and get 1000

Reply:
> from_str_radix doesn't support e/E so if you wrote 10E2 then you could treat it as a rational and get 1000.0 or treat the 10 and the 2 separately and get 1000

Let me try to explain this a bit better...

we have the `NumericKind` struct which contains the numeric kinds, if a number is for example:
 - `1000` => integer kind (can fit in an `i32`)
 - `0xfff` => integer kind (can fit in an `i32`)
 - `0xffffffffffffffff` => integer kind (can't fit in an `i32`, we store it as f64)
 - `1000000000000000000000000000` => integer kind (can't fit in an `i32`, we store it as f64)
 - `1000.0` => rational kind because it contains a `.`
 - `1000e10` => rational kind because it contains a `e`/`E`

as you can see in the example we can only get an integer kind if we can't get `e` (exponentianl notation) only when the kind is a `NumericKind::Ratioanl` so if we wanted to store as integer in cases where it can fit in an `i32` for example
 - `1000.0`
 - `100e2`
 - `1000.00`

these all can fit in an `i32` we need to in the `Rational` case:
```rust
NumericKind::Rational => {
	let number = f64::from_str(&buf).unwrap();
	if let Ok(integer) = i32::try_from(number) {
		// return integer because it can fit in i32
	}

	// return number because it cant fit in an i32
}
```

Reply:
I think I get what you are saying - so this would replace the below code?


                       let n = i32::from_str(&exp_str).map_err(|_| Error::syntax("Could not convert value to f64"))?;

                        if n < 0 { // A negative exponent is expected to produce a decimal value.

                            Numeric::Rational(

                                (num as f64) * f64::powi(10.0, n)

                            )

                        } else if let Some(exp) = i32::checked_pow(10, n as u32) {
                               if let Some(val) = i32::checked_mul(num, exp) {
                                   Numeric::Integer(val)
                               } else {
                                    Numeric::Rational(
                                        (num as f64) * (exp as f64)
                                    )
                               }
                        } else {
                            Numeric::Rational(
                                (num as f64) * f64::powi(10.0, n)
                            )
                        }`

Reply:
First we need to remove `exp_str` and append to `buf` where we appended to `exp_str`

In the `NumericKind::Rational` case:
```rust
NumericKind::Rational => {
	let number = f64::from_str(&buf).unwrap();
	if let Ok(integer) = i32::try_from(number) {
		// return integer because it can fit in i32
	}

	// return number because it cant fit in an i32
}
```
In the `NumericKind::Integer` case:
```rust
NumericKind::Integer(base) => {
	if let Ok(integer) = i32::from_str_radix(&buf, base) {
		// return integer;
		} else {
			let b = f64::from(base);
    		let mut result = 0.0_f64;
   			for c in buf.chars() {
        		let digit = f64::from(c.to_digit(base as u32));
    			result = result * b + digit;
    		}
			// return result
		}
	}
}
```

Reply:
I didn't find a definition for i32::try_from(num) where num is f64, is there something I'm missing?

Reply:
> I didn't find a definition for i32::try_from(num) where num is f64, is there something I'm missing?

Sorry I thought that it existed, in that case we can do something like:
```rust
NumericKind::Rational => {
	let number = f64::from_str(&buf).unwrap();
	if ((number as i32) as f64) == number {
		return number as i32 // because it can fit in i32
	}

	// return number because it cant fit in an i32
}
```
here we check if converting to `i32` does not change the value if it doesn't that means it can fit in a `i32`

Reply:
Ok this makes sense and the reasoning is because it makes the code cleaner by only requiring buf?

Reply:
> Ok this makes sense and the reasoning is because it makes the code cleaner by only requiring buf?

Yes. that is one of the reason but also it is more efficient than having two separate allocations

Reply:
> > Ok this makes sense and the reasoning is because it makes the code cleaner by only requiring buf?
> 
> Yes. that is one of the reason but also it is more efficient than having two separate allocations

ok cool, I shall make the changes once I've finished switching cursor to Result<Option<>> :)

Reply:
Done